R4R: CoreContext Refactor

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes. 
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  --> 

## ChangeLog
* Separated logic/concerns from `CoreContext`. `QueryContext` now facilitates querying and sending txs and `TxContext` facilitates building and signing txs.
* Cleaned up a bunch of the actual query context code
  * Separated into smaller composable methods with clear godocs
* Implemented `GetKeyInfo` and `GetPassphrase` as a general helper as I saw the same logic being used a lot
* Added enums for `key.Info` opposed to checking just strings

I've updated the `democoin` cool cmd to show how the new contexts would be used. One thing I came across and am not sure if it's the best approach is the `utils.SendTx`. That is what `EnsureSignBuildBroadcast` used to handle, but now that the logic and concerns have been split, not sure where to place that. Maybe `utils` is fine?

<hr>

* [ ] Updated all relevant documentation (`docs/`)
* [x] Updated all relevant code comments
* [ ] Wrote tests
* [x] Updated `CHANGELOG.md`
* [ ] Updated `cmd/gaia` and `examples/`

closes: #1551
___________________________________
For Admin Use:
* [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)
* [x] Reviewers Assigned 
* [x] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

ValarDragon commented on client/utils/utils.go:
```diff
@@ -0,0 +1,74 @@
+// signed transaction to a node.
+func SendTx(txCtx authctx.TxContext, queryCtx context.QueryContext, from []byte, name string, msgs []sdk.Msg) error {
+	if txCtx.AccountNumber == 0 {
+		accNum, err := queryCtx.GetAccountNumber(from)
```
When rebasing, can you be sure to add the "EnsureAccountExists" function?

alexanderbez replied:
@ValarDragon why is that? `SendTx` does what `EnsureAccountExists` virtually did. Do you want me to rip that out and call it `EnsureAccountExists` which `SendTx` would call?

alexanderbez replied:
I see what you mean. Yeah I think I can leverage something similar 👍

alexanderbez replied:
I'll add a method to `QueryContext#EnsureAccountExists`. This will get called before the any signing even happens.

alexanderbez commented:
@ValarDragon @cwgoes would love some initial/high-level feedback if this is a good way to approach things.

ValarDragon commented:
Anything else you're looking for feedback on, other than what I commented on?

alexanderbez commented:
@ValarDragon, nope! I updated the `cool` example. It now checks that the account exists first.

ValarDragon commented:
Maybe i'm misunderstanding. I thought `SendTx` is intended to replace the `EnsureSignBuildBroadcast` method. If so, then we want this check to go inside of SendTx similar to whats done here: https://github.com/cosmos/cosmos-sdk/blob/develop/client/context/viper.go#L76. I think we want this in the send functionality, not inside of everything trying to call send. (Similar to how we get account number and sequence number in send)

cwgoes commented on client/context/context.go:
```diff
@@ -0,0 +1,113 @@
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+)
+
+const ctxAccStoreName = "acc"
```
Hmm should this be a constant? Why not a `QueryContext` field in case we want to change it?

cwgoes replied:
In fact we have `QueryContext.AccountStore`

alexanderbez replied:
Yes, indeed it is already a field in the `QueryContext`, however:
1. There is no path (from what I can tell) where this is actually configurable or desirable (i.e. everywhere in the code uses `acc`)
2. Since there is no CLI flag for it, a user can just do `queryCtx.WithAccountStore(...)`.

I'm happy to revert it to being hard-coded.

jaekwon replied:
looks fine for now.

cwgoes commented on client/context/context.go:
```diff
@@ -0,0 +1,113 @@
+func (ctx QueryContext) WithUseLedger(useLedger bool) QueryContext {
+	ctx.UseLedger = useLedger
+	return ctx
+}
```
There are some fields in `QueryContext` without `WithXYZ` functions, although I guess if you didn't need to use them it doesn't matter.

alexanderbez replied:
Ok, I'll leave as is for now 👍

cwgoes commented on client/context/query.go:
```diff
@@ -0,0 +1,309 @@
+		return res, err
+	}
+
+	if res.CheckTx.Code != uint32(0) {
```
Can we reference some constant here - `sdk.ABCICodeOK` in `types/` IIRC?

cwgoes commented on client/context/query.go:
```diff
@@ -0,0 +1,309 @@
+			res.CheckTx.Log)
+	}
+
+	if res.DeliverTx.Code != uint32(0) {
```
Ditto on using a constant.

cwgoes commented on client/context/query.go:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	resp := result.Response
+	if resp.Code != uint32(0) {
```
`sdk.ABCICodeOK`

cwgoes commented on client/utils/utils.go:
```diff
@@ -0,0 +1,56 @@
+		return err
+	}
+
+	if txCtx.AccountNumber == 0 {
```
This really isn't the right check - what if the account number was intentionally tried as `0`? I think we want to check whether or not the user passed the `--account-number` flag at all.

alexanderbez replied:
Hmmm, it seems to be identical to the logic that currently exists:

https://github.com/cosmos/cosmos-sdk/blob/develop/client/context/viper.go#L105-L122

If the client does indeed intentionally pass `0`, then it'll look it up automatically -- kind of quirky, I understand.

If, as you're suggesting, we want to check if the client passed `0` and actually go with that, then'll be need to introduce the dependency of viper here. Thoughts?

ValarDragon replied:
Can't we just change the default value to like `-1` to mean automatic?

alexanderbez replied:
Yeah, but then the UX will be a little weird cause the CLI description will have the default in there of `-1`.

Let me think on this a bit more -- I can probably come up with a cleaner way outside the scope of `utils`.

ValarDragon replied:
Sounds good! I'd actually recommend just putting a "TODO: <problem>" above the relevant lines / moving this to a new issue (since this isn't a performance regression). We can do these UI `nice to haves` until later! (I doubt practically anyone will be doing --sequence=0 for the next month)

alexanderbez replied:
Actually, fair point! I'll do this :-)

cwgoes commented on client/utils/utils.go:
```diff
@@ -0,0 +1,56 @@
+		txCtx = txCtx.WithAccountNumber(accNum)
+	}
+
+	if txCtx.Sequence == 0 {
```
Same as above, with `--sequence` instead of `--account-number`.

alexanderbez replied:
See above.

cwgoes commented on crypto/keys/types.go:
```diff
@@ -44,10 +44,31 @@ type Keybase interface {
+
+// String implements the stringer interface for KeyType.
+func (kt KeyType) String() string {
+	keyTypes := map[KeyType]string{
```
This can be a constant? No need to recreate a map every time `.String()` is called.

cwgoes commented on client/context/query.go:
```diff
@@ -0,0 +1,309 @@
+			}
+
+			ctx.Logger.Write(bz)
+			io.WriteString(ctx.Logger, "\n")
```
Would it be better to use the logger here instead of just `io.WriteString` and `fmt.Printf` in various places? Although I guess we want machine-readable output and the logger adds extra info - maybe we could have a `json` logger and a `human-readable` logger though.

alexanderbez replied:
Hmmm not sure I follow. The `io.WriteString` is just a convenience method for writing a newline. I can easily just append a newline byte to `bz`. The actual logger here is `os.Stdout`. I also don't see any `fmt.Printf`s.

ValarDragon commented:
@alexanderbez while we're still reviewing, could you hold off on squashing? It makes it harder to to check if a fix was resolved as intended.

alexanderbez commented:
@ValarDragon sorry, I squashed already (to resolve rebasing). I won't squash from here on out until approved 👍

alexanderbez commented:
@cwgoes, @ValarDragon  I believe I addressed majority of your concerns and comments.

Regarding some aspects being rife with conditionals, if you could point them out, I'd be happy to improve! But generally I found there is little way around it just due to the nature of the way the context is used.

Regarding @sunnya97's proposed integration of keepers -- great idea, but let's address that in a separate PR that uses this as the basis 👍

jaekwon commented on client/context/query.go:
```diff
@@ -0,0 +1,309 @@
+// queryStore performs a query from a Tendermint node with the provided a store
+// name and path.
+func (ctx QueryContext) queryStore(key cmn.HexBytes, storeName, endPath string) ([]byte, error) {
+	path := fmt.Sprintf("/store/%s/%s", storeName, endPath)
```
maybe "store" here should likewise be parameterized

cwgoes replied:
That prefix (`/store` for store queries) is hardcoded in baseapp

jaekwon commented on client/keys/utils.go:
```diff
@@ -29,6 +29,54 @@ func GetKeyBase() (keys.Keybase, error) {
+
+	// we only need a passphrase for locally stored keys
+	if keyInfo.GetType() == keys.TypeLocal {
+		passphrase, err = ReadPassphraseFromStdin(name)
```
TODO: Make Issue to address this before launch

jaekwon replied:
(we shouldn't be passing around passphrases even in memory)

ValarDragon replied:
There already is an issue for this. #864

cwgoes replied:
How would we avoid storing the passphrase (transiently) in memory? We need to decrypt the encrypted keyfile.

alexanderbez replied:
@cwgoes brings up a good point. Warrants further discussion. I'll add a reference to #864 in the code tho

jaekwon commented on client/tx/broadcast.go:
```diff
@@ -13,7 +13,7 @@ type BroadcastTxBody struct {
 
 // BroadcastTx REST Handler
-func BroadcastTxRequestHandlerFn(ctx context.CoreContext) http.HandlerFunc {
+func BroadcastTxRequestHandlerFn(queryTx context.QueryContext) http.HandlerFunc {
```
queryTx?

alexanderbez replied:
errr typo. gracias.

jaekwon commented on client/tx/search.go:
```diff
@@ -53,19 +57,22 @@ func SearchTxCmd(cdc *wire.Codec) *cobra.Command {
 }
 
-func searchTxs(ctx context.CoreContext, cdc *wire.Codec, tags []string) ([]txInfo, error) {
+func searchTxs(queryTx context.QueryContext, cdc *wire.Codec, tags []string) ([]txInfo, error) {
```
queryTx?

jaekwon commented on crypto/keys/types.go:
```diff
@@ -44,10 +44,31 @@ type Keybase interface {
+type KeyType uint
+
+// Info KeyTypes
+const (
```
maybe KeyType should be a string?

jaekwon commented on x/auth/client/context/context.go:
```diff
@@ -0,0 +1,152 @@
+)
+
+// TxContext implements a transaction context created in SDK modules.
+type TxContext struct {
```
TxBuilder?

jaekwon commented on x/auth/client/context/utils.go:
```diff
@@ -0,0 +1,22 @@
+
+// defaultChainID returns the chain ID from the genesis file if present. An
+// error is returned if the file cannot be read or parsed.
+func defaultChainID() (string, error) {
```
seems kinda weird to define this here... should it be defined in tendermint/commands?

jaekwon replied:
if so maybe a NOTE is enough.

alexanderbez replied:
I'll add a note because I do not have enough context here. Honestly, I think we should remove it all together and require that the user always provide the correct `chainID`.

jaekwon commented on PENDING.md:
```diff
@@ -68,3 +69,14 @@ BUG FIXES
 *  \#1839 Fixed bug where intra-tx counter wasn't set correctly for genesis validators
 * [tests] \#1675 Fix non-deterministic `test_cover` 
+* [client] \#1551: Refactored `CoreContext`
+  * Renamed `CoreContext` to `QueryContext`
```
propose: QueryContext -> CliContext.  Since Query isn't the only thing we're doing there.

jaekwon commented:
Overall great refactor for splitting the piece into two.  Have a few items commented, like renaming QueryContext to CliContext.

alexanderbez commented:
Updated to reflect the latest PR review(s):
* Cleaned up faulty variable names
* Add TODOs with issue references
* Renamed `QueryContext` => `CLIContext`


/cc @jaekwon @cwgoes

alexanderbez commented:
@cwgoes @jaekwon can this be merged?

cwgoes commented:
> @cwgoes @jaekwon can this be merged?

Verily!
